tests/pytests/unit/fileserver/test_s3fs.py

```diff
@@ -1,19 +1,134 @@
+import os
+
+import boto3
 import pytest
+import yaml
+
+# moto must be imported before boto3
+from moto import mock_s3
 
 import salt.fileserver.s3fs as s3fs
+import salt.utils.s3
+from tests.support.mock import patch
 
 
 @pytest.fixture
-def configure_loader_modules(tmp_path):
+def bucket():
+    with mock_s3():
+        yield "mybucket"
+
+
+@pytest.fixture(scope="module")
+def aws_creds():
+    return {
+        "aws_access_key_id": "testing",
+        "aws_secret_access_key": "testing",
+        "aws_session_token": "testing",
+        "region_name": "us-east-1",
+    }
+
+
+@pytest.fixture(scope="function")
+def configure_loader_modules(tmp_path, bucket):
     opts = {
         "cachedir": tmp_path,
+        "s3.buckets": {"base": [bucket]},
+        "s3.location": "us-east-1",
+    }
+    utils = {"s3.query": salt.utils.s3.query}
+
+    return {s3fs: {"__opts__": opts, "__utils__": utils}}
+
+
+@pytest.fixture(scope="function")
+def s3(bucket, aws_creds):
+    conn = boto3.client("s3", **aws_creds)
+    conn.create_bucket(Bucket=bucket)
+    return conn
+
+
+def make_keys(bucket, conn, keys):
+    for key, data in keys.items():
+        conn.put_object(
+            Bucket=bucket,
+            Key=key,
+            Body=data["content"],
+        )
+
+
+def verify_cache(bucket, expected):
+    for key, data in expected.items():
+        correct_content = data["content"]
+        cache_file = s3fs._get_cached_file_name(bucket, "base", key)
+        assert os.path.exists(cache_file)
+
+        if correct_content is None:
+            continue
+
+        with salt.utils.files.fopen(cache_file) as f:
+            content = f.read()
+            assert correct_content == content
+
+
+# [HACK] to always invalidate the cache and re-download on every update
+@pytest.fixture(scope="module", autouse=True)
+def _patch_cache_expire():
+    with patch("salt.fileserver.s3fs.S3_CACHE_EXPIRE", -1):
+        yield
+
+
```

Comment:
```suggestion
```
See above for this hack still getting used

Reply:
Done